src/main/kotlin/com/pubnub/api/workers/SubscribeMessageProcessor.kt

```diff
@@ -201,34 +204,56 @@ internal class SubscribeMessageProcessor(
         )
     }
 
-    private fun processMessage(subscribeMessage: SubscribeMessage): JsonElement? {
+    private fun processMessage(subscribeMessage: SubscribeMessage): Pair<JsonElement?, PubNubError?> {
         val input = subscribeMessage.payload
 
-        val cryptoModule = pubnub.cryptoModule ?: return input
+        val cryptoModule = pubnub.cryptoModule ?: return input to null
         // if we do not have a crypto module, there is no way to process the node; let's return.
 
         // if the message couldn't possibly be encrypted in the first place, there is no way to process the node;
         // let's return.
         if (!subscribeMessage.supportsEncryption()) {
-            return input
+            return input to null
         }
+        input!!
 
-        val inputText = if (pubnub.mapper.isJsonObject(input!!) && pubnub.mapper.hasField(input, "pn_other")) {
-            pubnub.mapper.elementToString(input, "pn_other")
-        } else {
+        val inputText = if (pubnub.mapper.isJsonObject(input)) {
+            // property pn_other is used when we want to send encrypted Push Notification, not whole JSON object is encrypted but only value of pn_other property
+            if (pubnub.mapper.hasField(input, PN_OTHER)) {
+                // JSON with pn_other property indicates that this is encrypted Push Notification
+                pubnub.mapper.elementToString(input, PN_OTHER)
+            } else {
+                // plain JSON object indicates that this is not encrypted message
+                return input to logAndReturnDecryptionError()
+            }
+        } else if (input.isJsonPrimitive && input.asJsonPrimitive?.isString == true) {
+            // String may represent not encrypted string or encrypted data. We will check this when decrypting.
             pubnub.mapper.elementToString(input)
```

Comment:
If we decide not to use JsonElement.processHistoryMessage here. I would advise to write unit tests for this method similar to those from JsonElement.processHistoryMessage plus extra test to handles:
"else if (input.isJsonPrimitive && input.asJsonPrimitive?.isString == true)"

Reply:
this case is tested already both in JsonElement test and SubscribeMessageProcessorTest (all cases - including strings, arrays, objects, pn_other)